provd/internal/services/user/user.go

```diff
@@ -0,0 +1,293 @@
+// Package user implements the User gRPC service.
+package user
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+
+	"github.com/canonical/ubuntu-desktop-provision/provd/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/godbus/dbus/v5"
+)
+
+const (
+	accountsDBusName  string = "org.freedesktop.Accounts"
+	accountsDBusPath  string = "/org/freedesktop/Accounts"
+	hostnameDBusName  string = "org.freedesktop.hostname1"
+	hostnameDBusPath  string = "/org/freedesktop/hostname1"
+	accountsUserIface string = "org.freedesktop.Accounts.User"
+	hostnameIface     string = "org.freedesktop.hostname1"
+)
+
+type Caller interface {
+	Call(method string, flags dbus.Flags, args ...interface{}) *dbus.Call
+	GetProperty(p string) (dbus.Variant, error)
+}
+
+type UserObjectFactory interface {
+	GetUserObject(userObjectPath dbus.ObjectPath) Caller
+}
+
+type userObjectFactoryImpl struct {
+	conn *dbus.Conn
+}
+
+func (f *userObjectFactoryImpl) GetUserObject(userObjectPath dbus.ObjectPath) Caller {
+	userObject := f.conn.Object(accountsDBusName, userObjectPath)
+	return userObject
+}
+
+type options struct {
+	accounts    Caller
+	hostname    Caller
+	userFactory UserObjectFactory
+}
+
+type option func(*options)
+
+// Service is the implementation of the User module service.
+type Service struct {
+	proto.UnimplementedUserServiceServer
+	accounts    Caller
+	hostname    Caller
+	userFactory UserObjectFactory
+}
+
+// New retuns a new instance of the User service.
+func New(bus *dbus.Conn, args ...option) *Service {
+	accounts := bus.Object("org.freedesktop.Accounts", "/org/freedesktop/Accounts")
+	hostname := bus.Object("org.freedesktop.hostname1", "/org/freedesktop/hostname1")
+	userFactory := &userObjectFactoryImpl{conn: bus}
+
+	opts := options{
+		accounts:    accounts,
+		hostname:    hostname,
+		userFactory: userFactory,
+	}
+
+	// Apply given options
+	for _, f := range args {
+		f(&opts)
+	}
+
+	return &Service{
+		accounts:    opts.accounts,
+		hostname:    opts.hostname,
+		userFactory: opts.userFactory,
+	}
+}
+
+func hashPassword(password string) (string, error) {
+	// Generate a 16-byte salt
+	salt := make([]byte, 16)
+	if _, err := rand.Read(salt); err != nil {
+		return "", err
+	}
+
+	// Hash the password with the salt
+	hasher := sha512.New()
+	hasher.Write(salt)
+	hasher.Write([]byte(password))
+	hashedBytes := hasher.Sum(nil)
+	hashedPassword := base64.RawStdEncoding.EncodeToString(hashedBytes)
+
+	// Format the hash in the SHA-512 crypt format
+	saltStr := base64.RawStdEncoding.EncodeToString(salt)
+	return fmt.Sprintf("$6$%s$%s", saltStr, hashedPassword), nil
+}
```

Comment:
This seems to generate invalid password hashes. I think the issue is that the random 16-byte `salt` is base64 encoded, resulting in a `saltStr` longer than 16 bytes. As I understand it the length of the salt string (between the '$' symbols has to be between 8 and 16 (inclusive) characters long. So it's not 16 random bytes but 16 random characters (from [this set](https://github.com/maximeh/buildroot/blob/master/package/mkpasswd/mkpasswd.c#L74-L75)).
Sample outputs of `mkpasswd -m sha512crypt` with password 'foo':
```
$6$Cl3G4BBXctuRu98R$m0QWgf9doBLkBGryt8IAGpKgJDz0Jpls1R4fRAGdMnT7MT8a22ufkTYofct68FynDCOeYAr4UQiONYniKLnYa/
$6$tS0n2eZeF0IubeMY$n/X.zz7QHntZDjZqmYt86Oa/61w1GY3cKNX6R.G4MU.vn7DRv3bHiNK0CeZlGDE2H9AH4LnRTHmecPJyWpzvy/
$6$sXz.U7C4ktDZrytw$FfPdFu9pPQoWQ9oVXABtRb9ZXwPTiOs0TpKlRGbNx0kSqrpRAuIIllpFuHJu2ItyC.mGobuAnwo.3bnP/NqXO0
```